cmd/lotus-shed/chain_index.go

```diff
@@ -0,0 +1,222 @@
+		}
+
+		backfill := cctx.Bool("backfill")
+
+		// Results Tracking
+		logGood := cctx.Bool("log-good")
+
+		failedRPCs := 0
+		successfulBackfills := 0
+		successfulValidations := 0
+		successfulNullRounds := 0
+
+		startTime := time.Now()
+		_, _ = fmt.Fprintf(cctx.App.Writer, "%s starting chainindex validation; from epoch: %d; to epoch: %d; backfill: %t; log-good: %t\n", currentTimeString(),
+			fromEpoch, toEpoch, backfill, logGood)
+
+		totalEpochs := fromEpoch - toEpoch + 1
+		for epoch := fromEpoch; epoch >= toEpoch; epoch-- {
+			if ctx.Err() != nil {
+				return ctx.Err()
+			}
+
+			if (fromEpoch-epoch+1)%2880 == 0 || epoch == toEpoch {
+				progress := float64(fromEpoch-epoch+1) / float64(totalEpochs) * 100
+				elapsed := time.Since(startTime)
+				_, _ = fmt.Fprintf(cctx.App.ErrWriter, "%s -------- Chain index validation progress: %.2f%%; Time elapsed: %s\n",
+					currentTimeString(), progress, elapsed)
+			}
+
+			indexValidateResp, err := api.ChainValidateIndex(ctx, abi.ChainEpoch(epoch), backfill)
+			if err != nil {
+				tsKeyCid, err := tipsetKeyCid(ctx, abi.ChainEpoch(epoch), api)
+				if err != nil {
+					return fmt.Errorf("failed to get tipset key cid for epoch %d: %w", epoch, err)
+				}
+				_, _ = fmt.Fprintf(cctx.App.Writer, "%s ✗ Epoch %d (%s); failure: %s\n", currentTimeString(), epoch, tsKeyCid, err)
+				failedRPCs++
+				continue
+			}
+
+			if indexValidateResp.Backfilled {
+				successfulBackfills++
+			} else if indexValidateResp.IsNullRound {
+				successfulNullRounds++
+			} else {
+				successfulValidations++
+			}
+
+			if !logGood {
+				continue
+			}
+
+			if indexValidateResp.IsNullRound {
+				tsKeyCid, err := tipsetKeyCid(ctx, abi.ChainEpoch(epoch), api)
+				if err != nil {
+					return fmt.Errorf("failed to get tipset key cid for epoch %d: %w", epoch, err)
+				}
+				_, _ = fmt.Fprintf(cctx.App.Writer, "%s ✓ Epoch %d (%s); null round\n", currentTimeString(), epoch,
+					tsKeyCid)
+			} else {
+				jsonData, err := json.Marshal(indexValidateResp)
+				if err != nil {
+					return fmt.Errorf("failed to marshal results to JSON: %w", err)
+				}
+
+				_, _ = fmt.Fprintf(cctx.App.Writer, "%s ✓ Epoch %d (%s)\n", currentTimeString(), epoch, string(jsonData))
+			}
+		}
+
+		_, _ = fmt.Fprintf(cctx.App.Writer, "\n%s Chain index validation summary:\n", currentTimeString())
+		_, _ = fmt.Fprintf(cctx.App.Writer, "Total failed RPC calls: %d\n", failedRPCs)
+		_, _ = fmt.Fprintf(cctx.App.Writer, "Total successful backfills: %d\n", successfulBackfills)
+		_, _ = fmt.Fprintf(cctx.App.Writer, "Total successful validations without backfilling: %d\n", successfulValidations)
+		_, _ = fmt.Fprintf(cctx.App.Writer, "Total successful Null round validations: %d\n", successfulNullRounds)
+
+		return nil
+	},
+}
+
+func tipsetKeyCid(ctx context.Context, epoch abi.ChainEpoch, a api.FullNode) (cid.Cid, error) {
+	ts, err := a.ChainGetTipSetByHeight(ctx, epoch, types.EmptyTSK)
```

Comment:
this call is more expensive than it should be

current `indexes.go` just uses the fact that you can `.Parents()` on a tipset to get the parent one, super-efficient, no need to query the node for it. But here, you have an RPC query and you start at the head and work backward. Even with the skiplist this is not particualrly cheap.

Could you refactor to just use the `.Parents()` approach once you have the start epoch, or if that's too hard for some reason, can we pass the last known parent to the tipset argument of this call so it's cheaper?

Reply:
@rvagg

Actually, we were only using this function to get the tsKeyCid so we could log it. However, thinking through it, there is no need to hammer the API for info that the user can easily get from an explorer. _Maybe_ not the best UX but I'm okay with it to not do API calls that are not necessary. Have removed this alltogether.